File: test/dummy/app/components/dummy/ui/icon_component.rb
# frozen_string_literal: true

class Dummy::Ui::IconComponent < ApplicationComponent
  ICONS = YAML.load_file(::Rails.root.join("data/icons.yaml")).deep_symbolize_keys

  def initialize(name:, class_name: nil, width: nil, height: nil, top: nil, data: nil)
    @name = name
    raise "Unknown icon - #{@name}" if @name.blank? || !default_size

    @class_name = class_name
    @width = width
    @height = height
    @top = top
    @data = data
  end

  def default_size
    @default_size ||= ICONS[@name]
  end

  def class_names
    str = "d-ui-icon d-ui-icon--#{@name}"
    str += " #{@class_name}" if @class_name
    str
  end

  def style
    if @width
      width = option_size_to_str(@width)

      if @height
        height = option_size_to_str(@height)
      else
        height = "auto"
      end
    elsif @height
      width = "auto"
      height = option_size_to_str(@height)
    else
      width = "#{default_size[:width]}px"
      height = "#{default_size[:height]}px"
    end

    base = "width: #{width}; height: #{height}"

    if @top.is_a?(Numeric)
      "#{base}; position: relative; top: #{@top}px"
    else
      base
    end
  end

  def option_size_to_str(val)
    if val.is_a?(Numeric)
      "#{val}px"
    else
      val
    end
  end

  def js_default_sizes
    ICONS.to_json
  end
end

File: test/dummy/app/components/dummy/ui/icon_component.js
window.Dummy = window.Dummy || {}
window.Dummy.Ui = window.Dummy.Ui || {}
window.Dummy.Ui.Icon = window.Dummy.Ui.Icon || {}

window.Dummy.Ui.Icon.SVG_NS = 'http://www.w3.org/2000/svg'
window.Dummy.Ui.Icon.SVG_XLINK = 'http://www.w3.org/1999/xlink'

window.Dummy.Ui.Icon.data = (name, options = {}) => {
  if (!window.Dummy.Ui.Icon.svgSpritePath) throw new Error('Missing svgSpritePath. Add render(Dummy::Ui::IconComponent.new(head_html: true)) to <head>.')

  const data = { name }

  data.defaultSize = window.Dummy.Ui.Icon.defaultSizes[name]
  if (!data.defaultSize) throw new Error(`Unknown icon ${name}`)

  data.viewBox = `0 0 ${data.defaultSize.width} ${data.defaultSize.height}`

  data.style = {}

  if (options.height) {
    data.style.height = `${options.height}px`
    data.style.width = 'auto'
  } else if (options.width) {
    data.style.height = 'auto'
    data.style.width = `${options.width}px`
  } else {
    data.style.height = `${data.defaultSize.height}px`
    data.style.width = `${data.defaultSize.width}px`
  }

  data.classNames = ['d-ui-icon', `d-ui-icon--${name}`]

  if (options.class) {
    options.class.split(' ').forEach((className) => data.classNames.push(className))
  }

  data.href = `${window.Dummy.Ui.Icon.svgSpritePath}#${name}`

  data.data = options.data || {}

  return data
}

window.Dummy.Ui.Icon.create = (name, options = {}) => {
  const data = window.Dummy.Ui.Icon.data(name, options)

  const svg = document.createElementNS(window.Dummy.Ui.Icon.SVG_NS, 'svg')
  svg.setAttribute('viewBox', data.viewBox)

  Object.keys(data.style).forEach((key) => {
    svg.style[key] = data.style[key]
  })

  data.classNames.forEach((className) => {
    svg.classList.add(className)
  })

  Object.keys(data.data).forEach((key) => {
    svg.dataset[key] = data.data[key]
  })

  const use = document.createElementNS(window.Dummy.Ui.Icon.SVG_NS, 'use')
  use.setAttributeNS(window.Dummy.Ui.Icon.SVG_XLINK, 'xlink:href', data.href)

  svg.appendChild(use)

  return svg
}
